File: apps/estoque/src/actions/movimentacoes.ts
'use server'

import { z } from 'zod'
import { prisma } from '@/lib/prisma'
import { exigirSessao } from '@/lib/auth'
import { revalidarTelas } from '@/lib/revalidar'
import { saldoDoMaterial } from '@/queries/saldos'
import {
  MOVIMENTACAO, SINAL_MOVIMENTACAO, EXIGE_OBRA, ROTULO_MOVIMENTACAO, type TipoMovimentacao,
} from '@/lib/dominio/constantes'

export type Resultado<T = void> = { ok: true; dados: T } | { ok: false; erro: string }

/** "" vira undefined: campo opcional em branco não deve gravar string vazia. */
const opcional = z.string().trim().optional().transform((v) => (v ? v : undefined))

/** Data de calendário em meia-noite UTC — o mesmo referencial do resto do sistema. */
const dataCalendario = z
  .string()
  .trim()
  .regex(/^\d{4}-\d{2}-\d{2}$/, 'Data inválida.')
  .transform((v) => {
    const [a, m, d] = v.split('-').map(Number)
    return new Date(Date.UTC(a, m - 1, d))
  })

const esquema = z.object({
  materialId: z.string().trim().min(1, 'Selecione o material.'),
  tipo: z.enum([
    MOVIMENTACAO.ENTRADA, MOVIMENTACAO.SAIDA, MOVIMENTACAO.DEVOLUCAO,
    MOVIMENTACAO.PERDA, MOVIMENTACAO.AJUSTE_POSITIVO, MOVIMENTACAO.AJUSTE_NEGATIVO,
  ]),
  quantidade: z.coerce.number().positive('A quantidade tem de ser maior que zero.'),
  valorUnitario: z.union([z.coerce.number().nonnegative('Valor não pode ser negativo.'), z.literal('')]).optional(),
  obraId: opcional,
  fornecedorId: opcional,
  documento: opcional,
  observacao: opcional,
  ocorridoEm: dataCalendario,
})

export async function registrarMovimentacao(entrada: unknown): Promise<Resultado<{ id: string }>> {
  const sessao = await exigirSessao()

  const parsed = esquema.safeParse(entrada)
  if (!parsed.success) {
    return { ok: false, erro: parsed.error.issues.map((i) => i.message).join(' ') }
  }
  const d = parsed.data
  const tipo = d.tipo as TipoMovimentacao

  if (EXIGE_OBRA.includes(tipo) && !d.obraId) {
    return { ok: false, erro: `${ROTULO_MOVIMENTACAO[tipo]} precisa da obra de destino.` }
  }

  try {
    const material = await prisma.material.findUnique({
      where: { id: d.materialId },
      select: { id: true, nome: true, unidade: true, ativo: true },
    })
    if (!material) return { ok: false, erro: 'Material não encontrado.' }
    if (!material.ativo) return { ok: false, erro: 'Este material está inativo — reative o cadastro antes de movimentar.' }

    // Saída que deixaria o saldo negativo é recusada. Um saldo negativo não existe no
    // mundo físico: significa que a prateleira e o sistema discordam, e deixar passar
    // espalha esse desencontro por todo relatório daqui para frente. A mensagem aponta o
    // caminho certo — ajuste de inventário — em vez de só bloquear.
    if (SINAL_MOVIMENTACAO[tipo] === -1) {
      const saldo = await saldoDoMaterial(d.materialId)
      if (d.quantidade > saldo) {
        return {
          ok: false,
          erro:
            `Saldo insuficiente: há ${saldo} ${material.unidade} de "${material.nome}" em estoque. ` +
            'Se a quantidade física for outra, lance um ajuste de inventário antes.',
        }
      }
    }

    const movimentacao = await prisma.movimentacao.create({
      data: {
        materialId: d.materialId,
        tipo,
        quantidade: d.quantidade,
        valorUnitario: typeof d.valorUnitario === 'number' ? d.valorUnitario : null,
        // Obra e fornecedor só são gravados onde fazem sentido: uma perda não tem
        // fornecedor, e guardar um deixaria o extrato sugerindo que o fornecedor teve
        // alguma coisa a ver com a quebra.
        obraId: EXIGE_OBRA.includes(tipo) ? (d.obraId ?? null) : null,
        fornecedorId: tipo === MOVIMENTACAO.ENTRADA ? (d.fornecedorId ?? null) : null,
        documento: d.documento ?? null,
        observacao: d.observacao ?? null,
        ocorridoEm: d.ocorridoEm,
        registradoPor: sessao.nome,
      },
    })

    revalidarTelas('/', '/materiais', `/materiais/${d.materialId}`, '/movimentacoes', '/relatorios')
    return { ok: true, dados: { id: movimentacao.id } }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao registrar a movimentação.' }
  }
}

const esquemaInventario = z.object({
  materialId: z.string().trim().min(1, 'Selecione o material.'),
  quantidadeContada: z.coerce.number().nonnegative('A quantidade contada não pode ser negativa.'),
  observacao: opcional,
})

/**
 * Acerta o saldo pela contagem física.
 *
 * Quem conta a prateleira sabe quantas unidades tem, não a diferença para o sistema — pedir
 * a diferença obrigaria o almoxarife a fazer a conta de cabeça e a escolher o tipo certo,
 * duas chances de errar. Ele digita o que contou e o sistema deriva o ajuste.
 */
export async function ajustarPorInventario(entrada: unknown): Promise<Resultado<{ diferenca: number }>> {
  const sessao = await exigirSessao()

  const parsed = esquemaInventario.safeParse(entrada)
  if (!parsed.success) {
    return { ok: false, erro: parsed.error.issues.map((i) => i.message).join(' ') }
  }
  const d = parsed.data

  try {
    const material = await prisma.material.findUnique({ where: { id: d.materialId }, select: { id: true } })
    if (!material) return { ok: false, erro: 'Material não encontrado.' }

    const saldo = await saldoDoMaterial(d.materialId)
    const diferenca = d.quantidadeContada - saldo

    if (diferenca === 0) {
      return { ok: false, erro: 'A contagem bate com o saldo do sistema — nada a ajustar.' }
    }

    await prisma.movimentacao.create({
      data: {
        materialId: d.materialId,
        tipo: diferenca > 0 ? MOVIMENTACAO.AJUSTE_POSITIVO : MOVIMENTACAO.AJUSTE_NEGATIVO,
        quantidade: Math.abs(diferenca),
        ocorridoEm: new Date(),
        registradoPor: sessao.nome,
        observacao:
          `Inventário: contado ${d.quantidadeContada}, sistema apontava ${saldo}.` +
          (d.observacao ? ` ${d.observacao}` : ''),
      },
    })

    revalidarTelas('/', '/materiais', `/materiais/${d.materialId}`, '/movimentacoes', '/relatorios')
    return { ok: true, dados: { diferenca } }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao ajustar o inventário.' }
  }
}

File: migracao-go/internal/application/estoque/movimentacoes.go
package estoque

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"strings"
	"time"

	dominio "siqueiracampos/servidor/internal/domain/estoque"
	identidade "siqueiracampos/servidor/internal/domain/identidade"
)

type GerenciadorMovimentacoes struct {
	Materiais     dominio.MaterialRepositorio
	Movimentacoes dominio.MovimentacaoRepositorio
	Aprovacoes    dominio.AprovacaoRepositorio
	Configuracao  dominio.ConfiguracaoEmailRepositorio
	ClienteRH     dominio.ClienteRH
}

type EntradaMovimentacao struct {
	MaterialID, Tipo                                     string
	Quantidade, ValorUnitario                            string
	ObraID, FornecedorID, FuncionarioID, FuncionarioNome string
	Documento, Observacao, OcorridoEm                    string
}

type ResultadoMovimentacao struct {
	ID                string
	PendenteAprovacao bool
}

// Registrar espelha `registrarMovimentacao` de actions/movimentacoes.ts — a ordem das
// validações importa, ver COMPORTAMENTO.md §5.
func (g *GerenciadorMovimentacoes) Registrar(ctx context.Context, sess identidade.Sessao, e EntradaMovimentacao) (*ResultadoMovimentacao, error) {
	tipo := dominio.TipoMovimentacao(e.Tipo)
	if _, ok := dominio.RotuloMovimentacao[tipo]; !ok {
		return nil, erroValidacao("Tipo de movimentação inválido.")
	}
	materialID := strings.TrimSpace(e.MaterialID)
	if materialID == "" {
		return nil, erroValidacao("Selecione o material.")
	}
	quantidade := parseFloatOpcional(e.Quantidade)
	if quantidade <= 0 {
		return nil, erroValidacao("A quantidade tem de ser maior que zero.")
	}
	ocorridoEm := parseDataISO(e.OcorridoEm)
	if ocorridoEm == nil {
		return nil, erroValidacao("Data inválida.")
	}
	var valorUnitario *float64
	if v := strings.TrimSpace(e.ValorUnitario); v != "" {
		f := parseFloatOpcional(v)
		if f < 0 {
			return nil, erroValidacao("Valor não pode ser negativo.")
		}
		valorUnitario = &f
	}

	material, err := g.Materiais.BuscarPorID(ctx, materialID)
	if err != nil {
		return nil, err
	}
	if material == nil {
		return nil, erroValidacao("Material não encontrado.")
	}
	if !material.Ativo {
		return nil, erroValidacao("Este material está inativo — reative o cadastro antes de movimentar.")
	}

	// Quem é o destino depende da CATEGORIA, não só do tipo: EPI sai para uma pessoa, o
	// resto sai para uma obra — checar a categoria ANTES da obra, senão toda entrega de EPI
	// seria recusada por falta de obra que o formulário nem pede.
	precisaFuncionario := dominio.ExigeFuncionario(tipo, material.Categoria)
	var funcionarioID, funcionarioNome, obraID, fornecedorID *string

	if precisaFuncionario {
		funcionarioID = ponteiro(e.FuncionarioID)
		if funcionarioID == nil {
			return nil, erroValidacao("Saída de EPI precisa do funcionário que recebeu — é a ficha exigida pela NR-6.")
		}
		funcionarioNome = ponteiro(e.FuncionarioNome)
	} else if dominio.ExigeObra[tipo] {
		obraID = ponteiro(e.ObraID)
		if obraID == nil {
			return nil, erroValidacao(fmt.Sprintf("%s precisa da obra de destino.", dominio.RotuloMovimentacao[tipo]))
		}
	}
	if tipo == dominio.MovEntrada {
		fornecedorID = ponteiro(e.FornecedorID)
	}

	// Saída que deixaria o saldo negativo é recusada — não existe no mundo físico, e deixar
	// passar espalha o desencontro por todo relatório daqui pra frente.
	if dominio.SinalMovimentacao[tipo] == -1 {
		saldo, err := g.Materiais.SaldoDoMaterial(ctx, materialID)
		if err != nil {
			return nil, err
		}
		if quantidade > saldo {
			return nil, erroValidacao(fmt.Sprintf(
				"Saldo insuficiente: há %s %s de %q em estoque. Se a quantidade física for outra, lance um ajuste de inventário antes.",
				formatarNumero(saldo), material.Unidade, material.Nome))
		}
	}

	// Perda tira material do saldo sem nada entrando em troca. Acima de NENHUM limite: toda
	// perda lançada por quem não aprova passa pela gerência.
	if tipo == dominio.MovPerda && dominio.PrecisaAprovacao(identidade.PodeAprovar(sess.Cargo), true) {
		dados := dominio.DadosPerda{
			MaterialID: materialID, Quantidade: quantidade, OcorridoEm: ocorridoEm.UTC().Format(time.RFC3339),
			Documento: ponteiro(e.Documento), Observacao: ponteiro(e.Observacao),
		}
		b, err := json.Marshal(dados)
		if err != nil {
			return nil, err
		}
		aprovacao := &dominio.Aprovacao{
			Tipo: dominio.AprovacaoPerda, Dados: string(b),
			Resumo: fmt.Sprintf("%s: baixa de %s %s por perda ou quebra.", material.Nome, formatarNumero(quantidade), material.Unidade),
			Motivo: ponteiro(e.Observacao), SolicitanteID: sess.ID, SolicitanteNome: sess.Nome,
		}
		if err := g.Aprovacoes.Criar(ctx, aprovacao); err != nil {
			return nil, err
		}
		return &ResultadoMovimentacao{PendenteAprovacao: true}, nil
	}

	mov := &dominio.Movimentacao{
		Tipo: tipo, Quantidade: quantidade, ValorUnitario: valorUnitario,
		Documento: ponteiro(e.Documento), Observacao: ponteiro(e.Observacao), OcorridoEm: *ocorridoEm,
		RegistradoPor: ponteiro(sess.Nome), FuncionarioID: funcionarioID, FuncionarioNome: funcionarioNome,
		MaterialID: materialID, ObraID: obraID, FornecedorID: fornecedorID,
	}
	if err := g.Movimentacoes.Criar(ctx, mov); err != nil {
		return nil, err
	}

	// A saída já está gravada. Se o envio da ficha falhar, o material saiu do mesmo jeito —
	// a falha fica registrada na própria movimentação, nunca desfaz o que já aconteceu.
	if precisaFuncionario {
		g.sincronizarFicha(ctx, mov, material)
	}

	return &ResultadoMovimentacao{ID: mov.ID}, nil
}

func (g *GerenciadorMovimentacoes) sincronizarFicha(ctx context.Context, mov *dominio.Movimentacao, material *dominio.Material) {
	var validadeCA *string
	if material.ValidadeCA != nil {
		v := material.ValidadeCA.UTC().Format(time.RFC3339)
		validadeCA = &v
	}
	resultado := g.ClienteRH.EnviarFichaEpi(ctx, dominio.FichaEpi{
		MovimentacaoID: mov.ID, FuncionarioID: *mov.FuncionarioID,
		MaterialCodigo: material.Codigo, MaterialNome: material.Nome, Unidade: material.Unidade,
		Quantidade: mov.Quantidade, CA: material.CA, ValidadeCA: validadeCA,
		EntregueEm: mov.OcorridoEm.UTC().Format(time.RFC3339), EntreguePor: mov.RegistradoPor, Observacao: mov.Observacao,
	})
	if resultado.OK {
		_ = g.Movimentacoes.MarcarSincronizada(ctx, mov.ID, nil)
	} else {
		_ = g.Movimentacoes.MarcarSincronizada(ctx, mov.ID, &resultado.Erro)
	}
}

// SincronizarFicha reenvia (ou envia pela primeira vez) a ficha de uma saída de EPI —
// chamada tanto internamente quanto pelo botão de reenviar na tela. Nunca lança: erro vira
// texto pra tela mostrar, nunca exceção que faria a movimentação parecer que falhou.
func (g *GerenciadorMovimentacoes) SincronizarFicha(ctx context.Context, movimentacaoID string) error {
	mov, err := g.Movimentacoes.BuscarPorID(ctx, movimentacaoID)
	if err != nil {
		return err
	}
	if mov == nil {
		return erroValidacao("Movimentação não encontrada.")
	}
	if mov.FuncionarioID == nil {
		return erroValidacao("Esta movimentação não é uma entrega de EPI.")
	}
	if mov.SincronizadoEm != nil {
		return nil
	}
	material, err := g.Materiais.BuscarPorID(ctx, mov.MaterialID)
	if err != nil {
		return err
	}
	if material == nil {
		return erroValidacao("Material não encontrado.")
	}
	g.sincronizarFicha(ctx, mov, material)
	return nil
}

type EntradaAjuste struct {
	MaterialID, QuantidadeContada, Observacao string
}

type ResultadoAjuste struct {
	Diferenca         float64
	PendenteAprovacao bool
}

// AjustarPorInventario espelha `ajustarPorInventario` — a pessoa digita o que CONTOU, o
// sistema deriva o ajuste (pedir a diferença obrigaria conta de cabeça e escolha de tipo,
// duas chances de errar). Ver COMPORTAMENTO.md §5.
func (g *GerenciadorMovimentacoes) AjustarPorInventario(ctx context.Context, sess identidade.Sessao, e EntradaAjuste) (*ResultadoAjuste, error) {
	materialID := strings.TrimSpace(e.MaterialID)
	if materialID == "" {
		return nil, erroValidacao("Selecione o material.")
	}
	quantidadeContada := parseFloatOpcional(e.QuantidadeContada)
	if quantidadeContada < 0 {
		return nil, erroValidacao("A quantidade contada não pode ser negativa.")
	}

	material, err := g.Materiais.BuscarPorID(ctx, materialID)
	if err != nil {
		return nil, err
	}
	if material == nil {
		return nil, erroValidacao("Material não encontrado.")
	}

	saldo, err := g.Materiais.SaldoDoMaterial(ctx, materialID)
	if err != nil {
		return nil, err
	}
	diferenca := quantidadeContada - saldo
	if diferenca == 0 {
		return nil, erroValidacao("A contagem bate com o saldo do sistema — nada a ajustar.")
	}

	limites, err := obterLimites(ctx, g.Configuracao)
	if err != nil {
		return nil, err
	}
	if dominio.PrecisaAprovacao(identidade.PodeAprovar(sess.Cargo), math.Abs(diferenca) > limites.Ajuste) {
		sinal := "sobra"
		if diferenca < 0 {
			sinal = "falta"
		}
		dados := dominio.DadosAjusteInventario{
			MaterialID: materialID, QuantidadeContada: quantidadeContada, Observacao: ponteiro(e.Observacao),
		}
		b, err := json.Marshal(dados)
		if err != nil {
			return nil, err
		}
		aprovacao := &dominio.Aprovacao{
			Tipo: dominio.AprovacaoAjusteInventario, Dados: string(b),
			Resumo: fmt.Sprintf("%s: contagem de %s %s contra %s no sistema (%s de %s).",
				material.Nome, formatarNumero(quantidadeContada), material.Unidade, formatarNumero(saldo), sinal, formatarNumero(math.Abs(diferenca))),
			Motivo: ponteiro(e.Observacao), SolicitanteID: sess.ID, SolicitanteNome: sess.Nome,
		}
		if err := g.Aprovacoes.Criar(ctx, aprovacao); err != nil {
			return nil, err
		}
		return &ResultadoAjuste{Diferenca: diferenca, PendenteAprovacao: true}, nil
	}

	tipo := dominio.MovAjustePositivo
	if diferenca < 0 {
		tipo = dominio.MovAjusteNegativo
	}
	observacao := fmt.Sprintf("Inventário: contado %s, sistema apontava %s.", formatarNumero(quantidadeContada), formatarNumero(saldo))
	if strings.TrimSpace(e.Observacao) != "" {
		observacao += " " + strings.TrimSpace(e.Observacao)
	}
	mov := &dominio.Movimentacao{
		Tipo: tipo, Quantidade: math.Abs(diferenca), OcorridoEm: time.Now().UTC(),
		RegistradoPor: ponteiro(sess.Nome), Observacao: &observacao, MaterialID: materialID,
	}
	if err := g.Movimentacoes.Criar(ctx, mov); err != nil {
		return nil, err
	}
	return &ResultadoAjuste{Diferenca: diferenca}, nil
}
